### ISSUE TITLE
Return type for overloaded `__get__` at the descriptor is inferred incorrectly

### ISSUE THREAD

**Issue by saerhex** [astral-sh/ty#4084]
### Summary

Type inference works incorrectly for the overloaded getter method under the descriptor. See the simplified code example below:
```python
class Descriptor[**P, T]:
    @overload
    def __get__(self, instance: Proto[str], owner: Any) -> str:
        ...
    
    @overload
    def __get__(self, instance: Proto[int], owner: Any) -> int:
        ...
```
Assume we've assigned following descriptor. The inferred type for the attribute will always be the same one and it's the first overload definition. See the code below:

```python
class X[T](Proto[T]):
    foo = Descriptor()

x = X[int]()
```

[ty playground](https://play.ty.dev/e203f93f-f85b-4b7f-ae65-34f5882b945f)

### Version

0.0.63 (46f4915)

**Comment by carljm** [comment]
Thanks for the report! ty is actually correct here, but the reason is subtle and confusing. Your `Proto` and `X` types never use the `T` type variable at all. This causes ty's variance inference to (correctly) infer that `Proto` and `X` are bivariant in T -- in other words, `T` is entirely irrelevant and all specializations of `Proto/X` are equivalent to each other; there is no difference between `Proto[int]` and `Proto[str]`. Thus `X[int]` matches both `__get__` overloads, and (per the overload resolution rules) the first overload wins. Adding a use of `T` in the body of `Proto` [fixes the problem](https://play.ty.dev/430a894d-aa1c-4707-8779-10796bff6b1e).

We plan to remove bivariance (and fall back to covariance instead when bivariance is inferred) -- see #1728.
